### PR TITLE
Fix fsdp checkpoint strategy

### PR DESCRIPTION
**Pull request by bcui19**
# What does this PR do?

This PR allows us to now use method 2 for activation checkpointing. Basically FSDP recursively dives down to see if there are modules that should have activation checkpointing. However, we shouldn't be trying to wrap 'FullyShardedDataParallel' or 'FlattenedParamsWrappers' modules with the 'CheckpointWrapper'. By trying to wrap these two methods, it leads to a strange assert 

Ran with 7b param gpt-2 style models:
run with method 1: https://wandb.ai/mosaic-ml/fsdp-activation-checkpointing/runs/17oacy01
run with method 2: https://wandb.ai/mosaic-ml/fsdp-activation-checkpointing/runs/1f1ij64d

# What issue(s) does this change relate to?

Fixing [CO-1377](https://mosaicml.atlassian.net/jira/software/c/projects/CO/issues/CO-1377?jql=project%20%3D%20%22CO%22%20AND%20assignee%20%3D%20currentUser%28%29%20ORDER%20BY%20created%20DESC)

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/dist_strategy.py [hunk 1]
```diff
@@ -149,6 +149,18 @@ def prepare_fsdp_module(model: torch.nn.Module, optimizers: Optional[Union[torch
...
+    if optimizers:
+        optimizers_tuple = ensure_tuple(optimizers)
+        if len(optimizers_tuple) != 1:
+            raise NotImplementedError(f'Only one optimizer is supported; found {len(optimizers_tuple)} optimizers')
+
+        # clearing optimizer param groups and state
+        # that will be recreated at the end of prepare_fsdp_module
+        optim = optimizers_tuple[0]
+        optim.param_groups.clear()
+        optim.state.clear()
```
Comment (abhi-mosaic) [inline composer/trainer/dist_strategy.py:163]: This looks like a duplicate of the code right below it

File: composer/trainer/dist_strategy.py [hunk 2]
```diff
@@ -261,6 +273,8 @@ def _auto_wrap_policy(module: torch.nn.Module, recurse: bool, unwrapped_params:
                 # If module has attribute `module._activation_checkpointing = ...`, always respect it
                 # Otherwise checkpoint if root object `obj.activation_checkpointing_fn(module)` is true
                 def _check_fn(module: torch.nn.Module) -> bool:
+                    if isinstance(module, (FullyShardedDataParallel, FlattenParamsWrapper)):
+                        return False
```
Comment (abhi-mosaic) [inline composer/trainer/dist_strategy.py:277]: Trusting @bcui19 on tests, but this makes sense to me!

**Review by abhi-mosaic — APPROVED** [review]
Thank you for the quick fix! I think there is some duplicate code to cleanup, but approving now to unblock.

File: composer/trainer/dist_strategy.py (hunk 1 above)
Comment (bcui19) [reply composer/trainer/dist_strategy.py]: oops, had a bad merge

**Pull request merged**